core/modules/views/handlers/views_handler_field_bulk_form.inc

```diff
@@ -219,10 +219,15 @@ class views_handler_field_bulk_form extends views_handler_field {
   }
 
   /**
-   * Determine the base table of the bulk operation field.
+   * Determine the entity type of the bulk operation field.
    */
   protected function get_entity_type() {
-    $base_table = $this->view->base_table;
+    // Bulk operations are tied to a specific entity type that defines them.
+    // If the entity type definition is available, return it.
+    $table_data = views_fetch_data($this->table);
```

Comment:
It may be because I only tested with user and node, and because of my limited understanding of Views, but for me `$this->table` always returns the correct table for the chosen bulk handler.

You're absolutely right - `$this->view->base_table` is not correct here. But what about  `$this->table`? Wouldn't that alone be sufficient to get the right base_table?

Reply:
Let me see if I can answer your question. 

This PR solves the problem of the Bulk Form field not presenting the correct actions when we are dealing with a relationship. So, assume that you create a view that shows Content.  Your views base table is `node`. Assume you create a relationship to `user`, for example by adding "Content: Author" in the Relationship setting. Then assume you add a "User account" Bulk Operations field to the view. The current bug is that the handler (before you save it) will show you the actions for the views' base table entity (node) instead of the actions for the related entity (user). 

So, the fix basically consists in retrieving the correct entity for the handler.  What these lines are doing: they use `views_fetch_data` to retrieve the views data for the `user` table (which is the handler's table). The "views data" is what you defined in `user.views.inc` under the function `user_views_data()`. If you look at that function, you'll see that `$data['table']['entity type']` is set to `user`. So, this is an efficient, and foolproof way of getting the correct entity type from the handler table name.

Once you have the correct entity type, then you can retrieve the correct actions.

Let me know if that doesn't make sense...

Reply:
@argiepiano I see, I have to be clearer in my questions. :wink:

I'm aware that the PR fixes the bug and I'm also aware why it fixes it. But my finding is that $this->table in views_handler_field_bulk_form::get_entity_type() *always* seems to return the correct db table (user, node, comment). So it might be that the whole part in this function that tries to fetch additional data explicitly, if there's a relationship could be redundant.

The base table retrieved via $this->table seems to be sufficient (according to my experiments). My suspicion is that the function code could get simplified. But I might be overlooking something.

Reply:
Some more digging...

`$this->table` gets provided either by views_handler::init() or by views_handler_relationship::init(). That's why it's always the right table.

So the check could be: if `$this->table`, then the base table is `$this->table`, otherwise it's `$this->view->base_table`.

[This comment](https://github.com/backdrop/backdrop/blob/1.x/core/modules/views/includes/handlers.inc#L121) makes me think that `$this->table` might not always exist. So we should still fall back to $this->view->base_table.

But the fetching of additional data and also the extra check for options['relationship'] don't seem necessary. That would simplify the whole function.

Reply:
@indigoxela, it IS possible for the table and the entity to have different names, both for core entities and for custom entities. For example, try the following command:
dpm(views_fetch_data('taxonomy_term_data'));

In the case of taxonomy terms, the bulk form handler table would be defined as `taxonomy_term_data`, while the entity type is `taxonomy_term`.

The redundancy is there to prevent those cases. Do you disagree?

Reply:
> it IS possible for the table and the entity to have different names

Yes, of course. I'm not suggesting to only rely on the table name. That wouldn't even work for user (vs. users). :wink: 

My suggestion is to skip the extra fetch and fiddling with relationship data directly - that's unnecessary.

The part beginning with `foreach (entity_get_info() as $entity_type => $info) {` still *is* necessary, but the views_fetch_data() isn't and neither is `if (!empty($this->options['relationship']) ...`.

Reply:
Maybe thing are easier to understand, if I provide my suggestion as code. Explaining things can be hard. :wink: 

```
  protected function get_entity_type() {
    $base_table = isset($this->table) ? $this->table : $this->view->base_table;
    foreach (entity_get_info() as $entity_type => $info) {
      if (isset($info['base table']) && $info['base table'] == $base_table) {
        return $entity_type;
      }
      elseif (isset($info['revision table']) && $info['revision table'] == $base_table) {
        $this->revision = TRUE;
        return $entity_type;
      }
    }
    return FALSE;
  }
```
There still might be edge cases, where this wouldn't work as expected. But it's simpler and much easier to read and understand.

Reply:
@indigoxela the code above is make things clearer now.

OK, I'll make the change and then we can test. More later

Reply:
@indigoxela I made that change. Can you please test? Thanks